### ISSUE TITLE
V15 MDC Fixes

### ISSUE THREAD

**Issue by pedrojrivera** [johannesjo/angular-material-css-vars#117]
Here is a branch where I fixed up all the demo issues.  All contrast colors work as expected. Palette variables loading up properly now. This is a pure MDC implementation no legacy components. Previous legacy fixes have been removed.  I used a completely different strategy to determine when rgb or no-rgb variables should be used. 

https://github.com/pedrojrivera/angular-material-css-vars/tree/feature/v15-mdc

**Comment by pedrojrivera** [comment]
@johannesjo This branch doesn't consider legacy components, so there should be a distinction between this version and disclaim it. Anyone needing a legacy version will need to install an older release. That being said, I would be comfortable with this being merged back as it is. 

Just a note: not every Material component was tested, only the ones in this demo. I have a project that I work on where we use more components than the demo, so as I come across additional fixes I will add it to this project.

**Comment by johannesjo** [comment]
Thank you! Not sure what the best course of action is tbh, as I haven't made the switch to angular material 15 yet. Releasing multiple versions sounds like it might be confusing. I will revisit this subject once I do the migration myself or if someone else needs it (please comment below!).

**Comment by pedrojrivera** [comment]
So the good news is that we just finished converting a large application over to v15 pure mdc. So the final changes I committed are very complete in terms of a production worthy deployment. Everything is working as expected, including ripples, buttons, contrasting, etc.  I'll keep updating this branch if we find any new issues. 

I had a chat with another engineer and was suggesting that npm packages should probably be something like 

for legacy => angular-material-css-vars-legacy
mdc => angular-material-css-vars

The main reason is that eventually legacy will go away. Second, component structures are different between them. Finally, having both legacy and mdc together in the same project really bloats the application. 

Anyways.....these are just thoughts.

**Comment by ratatoeskr666** [comment]
@pedrojrivera I love your work! Can I merge parts of your scss overwrites into a fork of mine?

**Comment by pedrojrivera** [comment]
@ratatoeskr666  Absolutely...please let me know how it works out for you. If you encounter any issues let me know and I can look at them. Thanks!

**Comment by ratatoeskr666** [comment]
@pedrojrivera 
in the `_variables.scss` file there is a typo:
`900: var(--palette-warn-contrast-900b-no-rgb),`

**Comment by pedrojrivera** [comment]
@ratatoeskr666 Thanks. I'll have this fixed.

**Comment by pedrojrivera** [comment]
Pushed typo fixe to branch.

**Comment by jmkeppler** [comment]
Is there an npm for this version? I'm upgrading to v15 and @pedrojrivera's latest version on npm (4.2.2) has issues with getting the primary color using the public_utils.scss. Thanks!

**Comment by json-derulo** [comment]
@johannesjo any news on this? I have done the MDC migration by myself and encounter some issues now, which will be fixed by @pedrojrivera's changes.

I like the solution of having two separate npm packages. It should be relatively easy to set up, this repository just needs to be cloned. Then, the changes can be merged to this repository. Possibly my changes which allow both MDC and legacy to run simultaneously should be removed. Please let me know if you need any assistance.

**Comment by johannesjo** [comment]
@json-derulo still didn't need to do the migration myself on the only project I am using this on. If someone wants to release a separate package from a fork on their own I have nothing against it, but I won't have time for this myself. I am already doing a poor enough job of maintaining this project tbh. I feel sorry about it, but I just don't have enough time to give all my open source projects the love they deserve and this isn't likely to get better in the future :/

For this repo I am all for going down the route that requires the lowest effort on my part while assuring that the project remains useful to people. So for me it seems a good route to ditch any sort of backwards compatibility if it makes the process more complicated. People always can use the 14x versions.  

Not sure. What do you think?

**Comment by json-derulo** [comment]
@johannesjo One other option would be to give some people you trust write access to the repository and to npm. I would be willing to be a maintainer of this package.

I would also be fine with a new major version of this package which only supports MDC. If there is a need for the legacy version, someone else in the community can fork and maintain it.

**Comment by johannesjo** [comment]
That's great to hear @json-derulo ! I am adding you as co maintainer here. Let's discuss the npm stuff via email (contact@super-productivity.com). I would need your npm username.

**Comment by json-derulo** [comment]
@johannesjo thanks, I just sent you an email.

@pedrojrivera Seems like the plan is to release a new major version with full MDC and dropped legacy support. Could you create a pull request with your changes, so we can move forward?

**Comment by pedrojrivera** [comment]
@json-derulo Sorry I have been wrapped up in other work.  Did you still need anything from me? Thanks!

**Comment by json-derulo** [comment]
@pedrojrivera no worries. In the meantime, I have ported your commits into the `feature/mdc` branch of this repository. There were some nasty merge conflicts, I've fixed them. Would be great if you could have an extra look to the pull request #130 in case I have missed something.

**Comment by pedrojrivera** [comment]
@json-derulo Sure thing. I'll let you know what I find.

**Comment by pedrojrivera** [comment]
@json-derulo I checked your branch and reviewed the code. Built and tested. Everything is working as expected. The main/critical parts, which is the valriables.css and the service program calculations are all good. 

I noticed that the angular/material packages are a little behind.  I went ahead and updated all the libraries and retested. And everything works.  The latest material is at 15.2.8

Here is how I updated:

1. npm run update
    set typescrpt back to ~4.9.4
2. npm install --force

**Comment by json-derulo** [comment]
@pedrojrivera Thanks a lot! I will also take care of updating the dependencies.

I have one more question regarding your changes. When defining the themes in `_main.scss`, you removed `mat.define-dark-theme()` and `mat.define-dark-theme()`. As I can see, this leads to a whole bunch of issues, which were partially fixed by you afterwards. However I see new issues with background colors not applying, i.e. for disabled chips. I just tried adding back `mat.define-dark-theme()` and `mat.define-dark-theme()`, and now everything works like a charm, even without the workarounds you added. Can you remember why you did this change?

**Comment by pedrojrivera** [comment]
@json-derulo The material/angular libraries as they have been updated have forced me to make changes to work around certain issues, which very likely could then have been fixed and I never revisited some of those fixes to determine if they are actually needed. 

Go ahead and make your changes, I would be glad to pull and retest. The most important thing is to see if material is picking up the correct variables. Sometimes things seem "fixed" and it's just a coincidence. But material has had many internal fixes of issues I have had to account for in the past. These issues may very likely no longer exist. 

So I agree that some changes should be questioned to see if it is really needed anymore.

**Comment by json-derulo** [comment]
@pedrojrivera I have pushed the changes, everything looks good on my end. May I ask you to re-test?

**Comment by pedrojrivera** [comment]
@json-derulo Ok I did some tests and looking at your code I remembered a specific reason why I hade my palettes defined the way I had it.  (see my original code below). I had a project where I needed to have all the backgrounds be overwritten to be something other that white in light mode and another color in dark mode other than the material default.  I didn't want to override every single component with custom css when I should be able to provide a custom background palette for dark and light mode.  So your current change will not provide the ability to customize the background with variables as it will default to materials hard wired values internally.   So for example, the variable $mat-css-palette-background-dark is a copy of materials internal background palette. So if I want to globally change background colors for cards, dialogs, etc., I could just override it.  So in your version now the ability to override those particular variables is no longer possible. 

  $dark-theme: (
    color: (
      primary: $palette-primary,
      accent: $palette-accent,
      warn: $palette-warn,
      is-dark: true,
      foreground: $mat-css-palette-foreground-dark,
      background: $mat-css-palette-background-dark,
    ),
    typography: $typography-config,
    density: 0,
  );

And also 

// Background palette for dark themes.
$mat-css-palette-background-dark: (
  status-bar: rgb(var(--palette-background-status-bar-dark)),
  app-bar: rgb(var(--palette-background-app-bar-dark)),
  background: rgb(var(--palette-background-background-dark)),
  hover: rgba(var(--palette-background-hover-dark), var(--palette-background-hover-dark-alpha)),
  card: rgb(var(--palette-background-card-dark)),
  dialog: rgb(var(--palette-background-dialog-dark)),
  disabled-button: rgba(var(--palette-background-disabled-button-dark), var(--palette-background-disabled-button-dark-alpha)),
  raised-button: rgb(var(--palette-background-raised-button-dark)),
  focused-button: rgba(var(--palette-background-focused-button-dark), var(--palette-background-focused-button-dark-alpha)),
  selected-button: rgb(var(--palette-background-selected-button-dark)),
  selected-disabled-button: rgb(var(--palette-background-selected-disabled-button-dark)),
  disabled-button-toggle: rgb(var(--palette-background-disabled-button-toggle-dark)),
  unselected-chip: rgb(var(--palette-background-unselected-chip-dark)),
  disabled-list-option: rgb(var(--palette-background-disabled-list-option-dark)),
  tooltip: rgb(var(--palette-background-tooltip-dark)),
) !default;

**Comment by pedrojrivera** [comment]
@json-derulo Also, one more thing. Now when you look at this variable, $default-background and $default-foreground in _variables.scss, they are no longer meaningful since the variable they define are never used. 

When analyzing materials code, once the default background and foreground palettes have been defined, there is no way to overwrite them without initially passing in a custom background and foreground that use variables. If you don't pass in these custom palettes, internally they get created with defaults. 

So you could keep it this way, but know that all your variables for overriding the foreground and background are not being used.

**Comment by json-derulo** [comment]
@pedrojrivera In general, providing CSS variables also for background / foreground colors sounds like a reasonable feature to have. However, Angular Material's MDC-based components work with CSS variables. For instance, have a look at the background color of the card, it can be set via the `--mdc-elevated-card-container-color` variable:

<img width="985" alt="image" src="https://user-images.githubusercontent.com/18580672/233792232-80b7bc4c-2f53-4741-9a3a-3f5d02209b39.png">

In order to override the card background color at a central place you just need the following code:

```scss
.mat-mdc-card {
  --mdc-elevated-card-container-color: #000; // or whatever color you want to set it to
}
```

The problem I see with your implementation is that this package needs to maintain all possible background and foreground colors of all possible components. When Angular Material changes the colors, we need to manually adapt them here. This could be a quite big maintenance effort. If you find a way how to "automatically" pre-fill the values, I see no problem here.

Either way, I think this is more of a general change and not directly related to the MDC update. If you still want this feature, I am happy about your contribution!

**Comment by json-derulo** [comment]
> Also, one more thing. Now when you look at this variable, $default-background and $default-foreground in _variables.scss, they are no longer meaningful since the variable they define are never used.

I wonder what they are used for in the first place. Out of curiosity I have removed them. It seems like the demo application still works like before. Also, all the warnings like `var(--palette-primary-500-rgb) is not a color. Falling back to "dark" tone.` are gone.

@pedrojrivera I think you implemented those parts. Could you elaborate the benefits?

**Comment by pedrojrivera** [comment]
@json-derulo I was going to suggest that. But I just wanted to explain the logic since that logic was put in before they did their own variables and I was having to support legacy and non-legacy. I don't use legacy any more, so overriding their variables is the way to go. Also, many times that their implementation is inconsistent. So in some cases they have a variable you can override and other time you don't.    

The $default-background and $default-foreground, were just used as defaults to override material's internals. It was used to get variables in the css root. Then define $mat-css-palette-background-dark to use those variables, then passing them in as overriding default to material. 

> Also, all the warnings like var(--palette-primary-500-rgb) is not a color. Falling back to "dark" tone. are gone.

Material library fixed their code a while back to ignore variables, where a calculation was occurring to determine a foreground color based on whether the palette was dark or light mode, is-dark. In other places they were performing calculations assuming they were getting a legitimate hex value and not a variable. This would cause those waning messages. 

Also know that material returns values as RGB(), RGBA(), or just a hex values.  So when material calculates a value with alpha, they return the full expression as rgba(hexval, alpha).  So i needed to make sure that my overrides did not include the rgba function expressions in those cases and provide 2 variable to override.   

When no alpha is used, then they just return the actual hex value. Since I calculate all the values as rgb values, I need to return the full rgb() or rgba() expression. So at times you need the full rgba() and other times you don't.  

Not sure if you needed all this explanation but, this is what I had to deal with.  Internally, material handles the majority of their default values as hex values. This library deals with all the calculations as RGB/A values. So there has been a need to ensure that we play nice with material by having both flavors. 

Now, with all that said, one way to get rid of the -no-rgb variables, would be to override every possible internal material variable they use.

**Comment by pedrojrivera** [comment]
@json-derulo The program material-css-vars.ts is where the -no-rgb is calculated. You could comment out the function

private _replaceNoRgbValue(...)

This would give you a pretty good ide as to whether those variables are still needed with the latest material version.

**Comment by pedrojrivera** [comment]
@json-derulo Actually just comment out these lines in mat-css-color() to test

@function mat-css-color($hue: 500, $opacity: null, $palette: "primary", $is-contrast-color: false) {
  // If opacity is being provided, we need to to use *-no-rgb pallets
  // in order to return a proper variable value.
  //@if ($opacity != null) {
    //@return _mat-css-color-no-rgb($hue, $opacity, $palette, $is-contrast-color);
  //}

**Comment by json-derulo** [comment]
@pedrojrivera Thanks a lot for the detailed explanation!! I can definitely understand now much better what's going on.

As I just learned, overwriting background and foreground colors like card background has been supported since day one. However, I just cannot make it work for the legacy components (with the current state of the master branch). The colors will be directly set there is no variable which comes into action. I checked out tag v3.0.0, there it works like expected.

I wonder if it's a good idea in general to keep offering this to the users. We would need to add a lot of variables, and possibly there would be issues about wrong or missing colors (there have been some already in the past). According to the official Roadmap, Angular will start soon on token-based theming APIs, which could mean even more and frequent changes to the colors. Also like we already pointed out, the MDC-based components work with CSS variables, so for them it definitely seems redundant to offer additional variables.

Personally, I only need to be able to set primary, accent, warn colors, and toggle between light and dark mode at runtime. I would say this is the case for most of the users. At this moment I would feel most comfortable dropping support for foreground & background color variables, and only focus on adding variables for primary, accent and warn colors (+ possibly the existing text variables). Does anybody have an opinion on that, especially @johannesjo as an owner of the package?

**Comment by johannesjo** [comment]
Disclaimer: I am only superficially familiar with the code and the problems that come with the new mdc stuff.

tl;dr: Let's reduce it to the most important stuff!

Generally I think we should focus on keeping things simple to maintain und use. There are legacy versions for the legacy stuff if someone needs more control and sophisticated stuff. On top of that it sounds like there are already css variables available from mdc out of the box, which can be manipulated easily. Afaics it doesn't make sense to put another variable system with a different philosphy on top of that requires complicated hacks, unless there is a very very good reason for it (which it seems there isnt).

**Comment by json-derulo** [comment]
I just published version 5.0.0-next.0 which contains full MDC support. If you have any more issues with the MDC components, open a new issue. Please note that support for background and foreground color variables was dropped, as discussed in this thread.

I plan to create a stable release once Angular releases v16 and this package is updated to support it, which both should happen next week.
